locator/token_metadata.cc

```diff
@@ -773,122 +710,107 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
+            }
+            // apply new_normal_tokens
+            for (auto& [endpoint, tokens]: new_normal_tokens) {
+                result->update_topology(endpoint,
+                    get_dc_rack(endpoint),
+                    _normal_token_owners.contains(endpoint) ? node::state::normal : node::state::joining);
+                result->update_normal_tokens(std::move(tokens), endpoint).get();
+            }
+            // apply leaving endpoints
+            for (const auto& endpoint: _leaving_endpoints) {
+                result->remove_endpoint(endpoint);
+            }
+            result->sort_tokens();
+            return result;
+        }());
+
+        // merge tokens from token_to_endpoint and bootstrap_tokens,
+        // preserving tokens of leaving endpoints
+        const auto tokens = [&]() -> std::vector<dht::token> {
+            auto tokens = std::vector<dht::token>();
+            tokens.reserve(sorted_tokens().size() + get_bootstrap_tokens().size());
+            tokens.resize(sorted_tokens().size());
+            std::copy(begin(sorted_tokens()), end(sorted_tokens()), begin(tokens));
+            for (const auto& p: get_bootstrap_tokens()) {
+                tokens.push_back(p.first);
+            }
+            std::sort(begin(tokens), end(tokens));
+            return tokens;
+        }();
+
+        _pending_ranges_interval_map[keyspace_name] = [&]() -> ring_mapping {
+            ring_mapping pending_ranges;
+            for (size_t i = 0, size = tokens.size(); i < size; ++i) {
+                seastar::thread::maybe_yield();
+
+                const auto token = tokens[i];
+
+                const auto old_endpoints = strategy.calculate_natural_endpoints(token, unpimplified_this).get0();
+                const auto new_endpoints = strategy.calculate_natural_endpoints(token, new_token_metadata).get0();
+
+                auto add_mapping = [&](ring_mapping& target, std::unordered_set<inet_address>&& endpoints) {
+                    using interval = ring_mapping::interval_type;
+                    if (i == 0) {
+                        target += std::make_pair(
+                            interval::open(tokens.back(), dht::maximum_token()),
+                            endpoints);
+                        target += std::make_pair(
+                            interval::left_open(dht::minimum_token(), token),
+                            std::move(endpoints));
+                    } else {
+                        target += std::make_pair(
+                            interval::left_open(tokens[i - 1], token),
+                            std::move(endpoints));
+                    }
+                };
+
+                std::unordered_set<inet_address> pending_endpoints;
+                for (const auto& e: new_endpoints) {
+                    if (!old_endpoints.contains(e)) {
+                        pending_endpoints.insert(e);
+                    }
```

Comment:
Can insert unconditionally, since insert is idempotent.

Reply:
we need `pending_endpoints = new_endpoints.subtract(old_endpoints)`, can't do this unconditionally.

Reply:
I didn't even see that the set being tested isn't the set being updated.